src/core.storage.js

```diff
@@ -1,44 +1,106 @@
-class Storage {
-  static create(values, defaults) {
-    const store = new Storage();
-    for (const key of Object.keys(values)) {
-      store.setIfNull(values[key], defaults[key]);
+
+class ExtStore {
+  constructor() {
+    this._setLocal = function (obj) {
+      return new Promise(function (resolve) {
+        const entries = Object.entries(obj);
+
+        if (entries.length > 0) {
+          const [key, value] = entries[0];
+          window.localStorage.setItem(key, value);
+          resolve();
+        }
+      });
     }
-    return store;
-  }
 
-  set(key, val, cb) {
-    try {
-      localStorage.setItem(key, JSON.stringify(val));
-    } catch (e) {
-      const msg =
-        'Octotree cannot save its settings. ' +
-        'If the local storage for this domain is full, please clean it up and try again.';
-      console.error(msg, e);
+    this._getLocal = function (key) {
+      return new Promise(function (resolve) {
+        const value = window.localStorage.getItem(key);
+        resolve({[key]: value});
+      });
+    };
+
+    this._removeLocal = promisify(window.localStorage, 'removeItem');
+
+    // Safari hasn't supported storage at extension level yet. Fallback to localstorage
+    if (isSafari()) {
+      this._setAsync = this._setLocal;
+      this._getAsync = this._getLocal;
+      this._removeAsync = this._removeLocal;
+    } else {
+      this._setAsync = promisify(chrome.storage.local, 'set');
+      this._getAsync = promisify(chrome.storage.local, 'get');
+      this._removeAsync = promisify(chrome.storage.local, 'remove');
     }
-    if (cb) cb();
   }
 
-  get(key, cb) {
-    var val = parse(localStorage.getItem(key));
-    if (cb) cb(val);
-    else return val;
-
-    function parse(val) {
-      try {
-        return JSON.parse(val);
-      } catch (e) {
-        return val;
+  static create(values, defaults) {
+    return new Promise(async (resolve, reject) => {
+      const store = new ExtStore();
+      for (const key of Object.keys(values)) {
```

Comment:
May not have worthy perf impact, but generally I'd like to see this done in parallel, like

await Promise.all(keys.map((key) => store.setIfNull(...));